journalbeat/reader/journal.go

```diff
@@ -61,65 +60,63 @@ type Config struct {
 	logger  *logp.Logger
+	backoff *common.Backoff
 }
 
+// journalOpener is a function which tries to open the configured systemd journal.
+// If it is successful, a new logger instance is created also.
+type journalOpener func() (*sdjournal.Journal, *logp.Logger, error)
+
 // New creates a new journal reader and moves the FP to the configured position.
 func New(c Config, done chan struct{}, state checkpoint.JournalState, logger *logp.Logger) (*Reader, error) {
-	f, err := os.Stat(c.Path)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to open file")
-	}
-
-	logger = logger.With("path", c.Path)
-
-	var j *sdjournal.Journal
-	if f.IsDir() {
-		j, err = sdjournal.NewJournalFromDir(c.Path)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to open journal directory")
-		}
-	} else {
-		j, err = sdjournal.NewJournalFromFiles(c.Path)
+	opener := func() (*sdjournal.Journal, *logp.Logger, error) {
+		f, err := os.Stat(c.Path)
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to open journal file")
+			return nil, nil, errors.Wrap(err, "failed to open file")
 		}
-	}
 
-	err = setupMatches(j, c.Matches)
-	if err != nil {
-		return nil, err
-	}
-
-	r := &Reader{
-		journal: j,
-		changes: make(chan int),
-		config:  c,
-		done:    done,
-		logger:  logger,
+		var j *sdjournal.Journal
+		if f.IsDir() {
+			j, err = sdjournal.NewJournalFromDir(c.Path)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to open journal directory")
+			}
+		} else {
+			j, err = sdjournal.NewJournalFromFiles(c.Path)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to open journal file")
+			}
+		}
+		return j, logger.With("path", c.Path), nil
 	}
-	r.seek(state.Cursor)
-
-	instance.AddJournalToMonitor(c.Path, j)
 
-	r.logger.Debug("New journal is opened for reading")
-
-	return r, nil
+	return newReader(opener, c, done, state, logger)
 }
 
 // NewLocal creates a reader to read form the local journal and moves the FP
 // to the configured position.
 func NewLocal(c Config, done chan struct{}, state checkpoint.JournalState, logger *logp.Logger) (*Reader, error) {
-	j, err := sdjournal.NewJournal()
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to open local journal")
+	localOpener := func() (*sdjournal.Journal, *logp.Logger, error) {
+		j, err := sdjournal.NewJournal()
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to open local journal")
+		}
+
+		logger = logger.With("path", "local")
+		logger.Debug("New local journal is opened for reading")
+		return j, logger, nil
 	}
 
-	c.Path = LocalSystemJournalID
-	logger = logger.With("path", "local")
-	logger.Debug("New local journal is opened for reading")
+	return newReader(localOpener, c, done, state, logger)
+}
+
+func newReader(opener journalOpener, c Config, done chan struct{}, state checkpoint.JournalState, logger *logp.Logger) (*Reader, error) {
+	j, logger, err := opener()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
This function looks weird. We pass a logger as parameters, so to replace it with the logger instantiated via 'opener'?

The first `newReader` does is calling 'opener'. Let's simplify this by not passing a 'factory' here. Also, common context should come first:

```
func newReader(log *logp.Logger, done chan struct{}, cfg Config, journal *sdjournal.Journal, state checkpoint.JournalState) (*Reader, error) {
}
```